File: java/0112-path-sum.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean hasPathSum(TreeNode root, int targetSum) {
        return dfs(root, targetSum);
    }

    public boolean dfs(TreeNode node, int remainder) {
        if (node == null)
            return false;
        remainder -= node.val;
        if (remainder == 0 && node.left == null && node.right == null)
            return true;
        return dfs(node.left, remainder) || dfs(node.right, remainder);
    }
}

File: python/0112-path-sum.py
from typing import Optional

class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def hasPathSum(self, root: Optional[TreeNode], targetSum: int) -> bool:

        def dfs(node, remainder):
            if not node:
                return False
            remainder = remainder - node.val
            if remainder == 0 and not node.left and not node.right:
                return True
            return dfs(node.left, remainder) or dfs(node.right, remainder)
        
        return bool(dfs(root, targetSum))

        